mainapp/views.py

```diff
@@ -226,15 +229,26 @@ class Maintenance(TemplateView):
 
 def mapdata(request):
     district = request.GET.get("district", "all")
-    data = cache.get("mapdata:" + district)
-    if data:
-        return JsonResponse(list(data) , safe=False)
-    if district != "all":
-        data = Request.objects.exclude(latlng__exact="").filter(district=district).values()
-    else:
-        data = Request.objects.exclude(latlng__exact="").values()
-    cache.set("mapdata:" + district, data, settings.CACHE_TIMEOUT)
-    return JsonResponse(list(data) , safe=False)
+    page = max(int(request.GET.get("page", "1")), 1)
+
+    cachekey = "mapdata:" + district + ',page:' + str(page)
+    data = cache.get(cachekey)
+    if not data:
+        db = connection.cursor()
+        where, args = '', []
+
+        if district != "all":
+            where += 'where district = %s'
+            args.append(district)
+
+        args.extend([PER_PAGE, (page - 1) * PER_PAGE])
+        db.execute("SELECT * FROM mainapp_request {WHERE} ORDER BY id "
+                   "DESC LIMIT %s OFFSET %s;".format(WHERE=where), args)
```

Comment:
if you are using desc, the returned data won't be consistent as new entries are added. 

As you move from page 1 to 2, you might miss new entries that have been added while you where on page 1.

Reply:
I had considered it, but the frequent use case being fetching latest requests, the over all load on the server lower. Duplicates is tradeoff here. Looking at the last dump, we request peaked ~200 hour. Any suggestions ?

For now the syncing algo must be to keep fetching till last known ID. As IDs are series, we can figure out of items were missed.

Reply:
@naharp 

If the tradeoff's are acceptable then this looks fine. If people are using this to get via an API, It's okay. 

On a webpage the longer they stay on one page the less consistent the data becomes. I don't think we have that information, so people can just refresh this. That is okay.

The ideal way to do this would be use a token(last returned id) as well. This points to the last previous record that was fetched. If this token is provided use that instead of offset, else just use offset.